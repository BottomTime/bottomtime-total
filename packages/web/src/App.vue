<template>
  <div v-if="isLoading"></div>
  <div v-else>
    <SnackBar />
    <NavBar />
    <RouterView />
    <PageFooter />
    <CookieWarning />
  </div>
</template>

<script lang="ts" setup>
import { onBeforeMount, ref } from 'vue';

import CookieWarning from './components/main/CookieWarning.vue';
import { Dispatch, useStore } from '@/store';
import { inject } from '@/helpers';
import NavBar from '@/components/main/NavBar.vue';
import PageFooter from '@/components/main/PageFooter.vue';
import SnackBar from '@/components/main/SnackBar.vue';
import { ApiClientKey } from '@/injection-keys';

const client = inject(ApiClientKey);
const store = useStore();
const isLoading = ref(true);

onBeforeMount(async () => {
  try {
    const currentUser = await client.users.getCurrentUser();
    await store.dispatch(Dispatch.SignInUser, currentUser);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
});
</script>

<style>
@import './assets/bulmaswatch.min.css';
</style>
