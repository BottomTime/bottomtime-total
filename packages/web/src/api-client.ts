import { ApiClient } from '@bottomtime/api';

import { InjectionKey, inject } from 'vue';

export const ApiClientKey: InjectionKey<ApiClient> = Symbol('ApiClient');
export function useClient(): ApiClient {
  const client = inject(ApiClientKey);

  if (!client) {
    throw new Error(
      'Api Client has not been initialized. Did you forget to call `app.provide(key, client)`?',
    );
  }

  return client;
}
