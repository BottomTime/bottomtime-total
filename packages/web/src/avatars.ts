import { AvatarSize, ListAvatarURLsResponseDTO } from '@bottomtime/api';

export type AvatarURLs = ListAvatarURLsResponseDTO['sizes'];

export function getAvatarURL(
  baseUrl: string | undefined,
  size: AvatarSize,
): string | undefined {
  if (!baseUrl) return undefined;

  if (/^\/api\/users\/[a-z0-9]+([_.-][a-z0-9]+)*\/avatar/i.test(baseUrl)) {
    return baseUrl.endsWith('/') ? `${baseUrl}${size}` : `${baseUrl}/${size}`;
  }

  return baseUrl;
}
