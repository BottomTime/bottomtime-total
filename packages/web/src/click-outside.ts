import { Directive, DirectiveBinding } from 'vue';

type HTMLElementWithClickOutside = HTMLElement & {
  clickOutsideEvent: (e: MouseEvent) => void;
};

export const clickOutside: Directive = {
  beforeMount: (el: HTMLElementWithClickOutside, binding: DirectiveBinding) => {
    el.clickOutsideEvent = (event) => {
      // here I check that click was outside the el and his children
      if (!(el == event.target || el.contains(event.target as Node))) {
        // and if it did, call method provided in attribute value
        binding.value();
      }
    };
    document.addEventListener('click', el.clickOutsideEvent);
  },
  unmounted: (el: HTMLElementWithClickOutside) => {
    document.removeEventListener('click', el.clickOutsideEvent);
  },
};
