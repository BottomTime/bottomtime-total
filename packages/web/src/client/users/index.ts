export * from './interfaces';

export * from './default-user';
export * from './default-user-manager';
