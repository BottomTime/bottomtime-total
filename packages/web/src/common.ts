import { AccountTier, LogBookSharing, ProfileDTO } from '@bottomtime/api';

import { ComputedRef } from 'vue';

export type Breadcrumb = {
  label: string | ComputedRef<string>;
  to?: string | ComputedRef<string>;
  active?: boolean;
};

export type Coordinates = {
  left: number;
  top: number;
  width: number;
  height: number;
};

export const DateTimeFormat = 'yyyy-MMM-dd hh:mm:ss a';

export const DefaultProfile: ProfileDTO = {
  accountTier: AccountTier.Basic,
  logBookSharing: LogBookSharing.Private,
  memberSince: 0,
  username: '',
  userId: '',
};

export type Selectable<T> = T & { selected?: boolean };

export type SelectOption = {
  disabled?: boolean;
  label?: string;
  value: string;
};

export type TabInfo = {
  key: string;
  label: string;
  disabled?: boolean;
  visible?: boolean;
};

export enum ToastType {
  Error = 'error',
  Info = 'info',
  Success = 'success',
  Warning = 'warning',
}

export type Toast = {
  id: string;
  message: string;
  type: ToastType;
};

export const FortyEightHoursInSeconds = 48 * 60 * 60 * 1000;
