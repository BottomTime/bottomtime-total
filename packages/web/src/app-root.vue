<template>
  <SnackBar />
  <div
    id="modal"
    class="font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950"
  ></div>
  <div id="drawer"></div>

  <div
    class="md:container mx-auto font-content text-grey-900 bg-blue-200 dark:text-grey-100 dark:bg-blue-950 rounded-b-xl shadow-md shadow-white dark:shadow-grey-700 opacity-95"
  >
    <!-- Placeholder element. Modal dialogs will be "teleported" here to get the z-indexing right. -->

    <section class="mt-16 p-4">
      <RouterView></RouterView>
    </section>
    <PageFooter />
  </div>
  <NavBar />
</template>

<script setup lang="ts">
import { onBeforeMount } from 'vue';

import { ToastType } from './common';
import NavBar from './components/core/nav-bar.vue';
import PageFooter from './components/core/page-footer.vue';
import SnackBar from './components/core/snack-bar.vue';
import { Config } from './config';
import { useErrors, useToasts } from './store';

const toasts = useToasts();
const errors = useErrors();

onBeforeMount(() => {
  if (!Config.isSSR) {
    if (errors.renderError) {
      if (!Config.isProduction) {
        /* eslint-disable-next-line no-console */
        console.error(errors.renderError);
      }
      toasts.toast({
        id: 'server-prerender-error',
        message:
          'A server error has occurred and the page may not have loaded correctly. We apologize for the inconvenience. Please try again later as we investigate the issue.',
        type: ToastType.Error,
      });
    }
  }
});
</script>
