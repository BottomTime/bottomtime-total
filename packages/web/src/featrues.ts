import { Feature } from '@bottomtime/common';

import { User as ConfigCatUser, IConfigCatClient } from 'configcat-common';
import { InjectionKey, Reactive, inject, reactive } from 'vue';

import { useLogger } from './logger';
import { useCurrentUser } from './store';

const log = useLogger('features');

export const FeaturesServiceKey: InjectionKey<IConfigCatClient> =
  Symbol('FeaturesService');

export type FeatureValue = Reactive<{
  isLoading: boolean;
  value: boolean;
}>;

export function useFeatureToggle(feature: Feature<boolean>): FeatureValue {
  const instance = reactive<FeatureValue>({
    isLoading: true,
    value: feature.defaultValue,
  });

  const currentUser = useCurrentUser();
  const client = inject(FeaturesServiceKey);
  if (!client) {
    throw new Error(
      'FeaturesService has not been initialized. Did you remember to call `app.provide(key, client)`?',
    );
  }

  const user: ConfigCatUser | undefined = currentUser.user
    ? new ConfigCatUser(currentUser.user.id, currentUser.user.email)
    : undefined;

  client
    .getValueAsync(feature.key, feature.defaultValue, user)
    .then((flag) => {
      instance.value = flag;
    })
    .catch((err) => {
      log.error(err);
    })
    .finally(() => {
      instance.isLoading = false;
    });

  client.on('configChanged', (newConfig) => {
    const newValue = newConfig.settings[feature.key]?.value;
    if (newValue !== undefined) {
      instance.value = newValue as boolean;
    }
  });

  return instance;
}
