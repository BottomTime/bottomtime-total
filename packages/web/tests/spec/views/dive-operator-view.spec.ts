import {
  AccountTier,
  ApiClient,
  CreateOrUpdateDiveOperatorDTO,
  DiveOperator,
  DiveOperatorDTO,
  Fetcher,
  HttpException,
  UserDTO,
  VerificationStatus,
} from '@bottomtime/api';
import { ManageDiveOperatorsFeature } from '@bottomtime/common';

import {
  ComponentMountingOptions,
  flushPromises,
  mount,
  renderToString,
} from '@vue/test-utils';

import { Pinia, createPinia } from 'pinia';
import { Router } from 'vue-router';

import { ApiClientKey } from '../../../src/api-client';
import { ToastType } from '../../../src/common';
import EditDiveOperator from '../../../src/components/operators/edit-dive-operator.vue';
import ViewDiveOperator from '../../../src/components/operators/view-dive-operator.vue';
import { FeaturesServiceKey } from '../../../src/featrues';
import { LocationKey, MockLocation } from '../../../src/location';
import {
  useCurrentUser,
  useDiveOperators,
  useToasts,
} from '../../../src/store';
import DiveOperatorView from '../../../src/views/dive-operator-view.vue';
import { ConfigCatClientMock } from '../../config-cat-client-mock';
import { createRouter } from '../../fixtures/create-router';
import {
  BlankDiveOperator,
  FullDiveOperator,
  PartialDiveOperator,
} from '../../fixtures/dive-operators';
import { AdminUser, BasicUser } from '../../fixtures/users';

const ShopOwner: UserDTO = {
  ...BasicUser,
  accountTier: AccountTier.ShopOwner,
  profile: {
    ...BasicUser.profile,
    accountTier: AccountTier.ShopOwner,
  },
};

describe('Dive Operator view', () => {
  let fetcher: Fetcher;
  let client: ApiClient;
  let features: ConfigCatClientMock;
  let location: MockLocation;
  let router: Router;

  let pinia: Pinia;
  let currentUser: ReturnType<typeof useCurrentUser>;
  let diveOperators: ReturnType<typeof useDiveOperators>;
  let toasts: ReturnType<typeof useToasts>;
  let opts: ComponentMountingOptions<typeof DiveOperatorView>;

  beforeAll(() => {
    fetcher = new Fetcher();
    client = new ApiClient({ fetcher });
    features = new ConfigCatClientMock();
    location = new MockLocation();
    router = createRouter([
      {
        path: '/shops/createNew',
        component: DiveOperatorView,
      },
      {
        path: '/shops/:shopKey',
        component: DiveOperatorView,
      },
    ]);
  });

  beforeEach(() => {
    pinia = createPinia();
    currentUser = useCurrentUser(pinia);
    diveOperators = useDiveOperators(pinia);
    toasts = useToasts(pinia);
    features.flags[ManageDiveOperatorsFeature.key] = true;
    opts = {
      global: {
        plugins: [pinia, router],
        provide: {
          [ApiClientKey as symbol]: client,
          [FeaturesServiceKey as symbol]: features,
          [LocationKey as symbol]: location,
        },
        stubs: {
          teleport: true,
        },
      },
    };
  });

  describe('when rendering on the server side', () => {
    it('will render not found page if feature flag is off', async () => {
      features.flags[ManageDiveOperatorsFeature.key] = false;

      const div = document.createElement('div');
      div.innerHTML = await renderToString(DiveOperatorView, {
        global: opts.global,
      });

      expect(
        div.querySelector('[data-testid="view-operator-section"]'),
      ).toBeNull();
      expect(
        div.querySelector('[data-testid="not-found-message"]'),
      ).not.toBeNull();
    });

    it('will render not found page if operator is not found', async () => {
      await router.push('/shops/unknown-operator');

      const spy = jest
        .spyOn(client.diveOperators, 'getDiveOperator')
        .mockRejectedValue(
          new HttpException(404, 'Not found', 'Not found', {
            message: 'Not found',
            method: 'GET',
            path: '/dive-operators/unknown-operator',
            status: 404,
          }),
        );

      const div = document.createElement('div');
      div.innerHTML = await renderToString(DiveOperatorView, {
        global: opts.global,
      });

      expect(spy).toHaveBeenCalledWith('unknown-operator');
      expect(
        div.querySelector('[data-testid="view-operator-section"]'),
      ).toBeNull();
      expect(
        div.querySelector('[data-testid="not-found-message"]'),
      ).not.toBeNull();
    });

    it('will render dive operator details', async () => {
      await router.push(`/shops/${FullDiveOperator.slug}`);

      const spy = jest
        .spyOn(client.diveOperators, 'getDiveOperator')
        .mockResolvedValue(new DiveOperator(fetcher, FullDiveOperator));

      const div = document.createElement('div');
      div.innerHTML = await renderToString(DiveOperatorView, {
        global: opts.global,
      });

      expect(spy).toHaveBeenCalledWith(FullDiveOperator.slug);

      const html = div.querySelector('[data-testid="view-operator-section"]');
      expect(html).not.toBeNull();
      expect(html?.innerHTML).toMatchSnapshot();
    });

    it('will render in edit mode when the user is the shop owner', async () => {
      const operator: DiveOperatorDTO = {
        ...FullDiveOperator,
        owner: ShopOwner.profile,
      };
      currentUser.user = ShopOwner;
      await router.push(`/shops/${operator.slug}`);

      const spy = jest
        .spyOn(client.diveOperators, 'getDiveOperator')
        .mockResolvedValue(new DiveOperator(fetcher, operator));

      const div = document.createElement('div');
      div.innerHTML = await renderToString(DiveOperatorView, {
        global: opts.global,
      });

      expect(
        div.querySelector<HTMLInputElement>('input#operator-name')?.value,
      ).toBe(operator.name);
      expect(spy).toHaveBeenCalledWith(operator.slug);
    });
  });

  describe('when rendering on client side', () => {
    const update: CreateOrUpdateDiveOperatorDTO = {
      active: true,
      address: '1234 Main St',
      description: 'A new dive shop',
      name: 'New Shop',
      phone: '555-555-5555',
      slug: FullDiveOperator.slug,
      email: FullDiveOperator.email,
      gps: FullDiveOperator.gps,
      socials: FullDiveOperator.socials,
      website: FullDiveOperator.website,
    };

    const create: CreateOrUpdateDiveOperatorDTO = {
      active: true,
      address: '74 Main St',
      description: 'We do scuba stuff',
      name: 'Scuba Mega Shop',
      slug: 'scuba-mega-shop',
      phone: '555-555-5555',
      email: 'admin@scubamegashop.com',
      website: 'https://scubamegashop.com',
    };

    it('will allow user to view a dive operator', async () => {
      diveOperators.currentDiveOperator = { ...PartialDiveOperator };
      await router.push(`/shops/${PartialDiveOperator.slug}`);

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      const viewer = wrapper.getComponent(ViewDiveOperator);
      expect(viewer.isVisible()).toBe(true);
      expect(viewer.props('operator')).toEqual(PartialDiveOperator);
    });

    it('will show not found if operator is not found', async () => {
      diveOperators.currentDiveOperator = null;
      await router.push('/shops/unknown-operator');

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      expect(
        wrapper.find('[data-testid="not-found-message"]').isVisible(),
      ).toBe(true);
      expect(wrapper.findComponent(ViewDiveOperator).exists()).toBe(false);
    });

    it('will show login form if unauthenticated user tries to create a new operator', async () => {
      currentUser.user = null;
      diveOperators.currentDiveOperator = {
        ...BlankDiveOperator,
        owner: ShopOwner.profile,
      };
      await router.push('/shops/createNew');

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      expect(
        wrapper.get('[data-testid="require-auth-anonymous"]').isVisible(),
      ).toBe(true);
      expect(wrapper.findComponent(EditDiveOperator).exists()).toBe(false);
    });

    it('will show forbidden message when a regular user tries to create a new operator', async () => {
      currentUser.user = {
        ...BasicUser,
        accountTier: AccountTier.Basic,
      };
      diveOperators.currentDiveOperator = {
        ...BlankDiveOperator,
        owner: BasicUser.profile,
      };

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      expect(
        wrapper.find('[data-testid="forbidden-message"]').isVisible(),
      ).toBe(true);
      expect(wrapper.findComponent(EditDiveOperator).exists()).toBe(false);
    });

    it('will show forbidden message when a pro user tries to create a new operator', async () => {
      currentUser.user = {
        ...BasicUser,
        accountTier: AccountTier.Pro,
      };
      diveOperators.currentDiveOperator = {
        ...BlankDiveOperator,
        owner: BasicUser.profile,
      };

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      expect(
        wrapper.find('[data-testid="forbidden-message"]').isVisible(),
      ).toBe(true);
      expect(wrapper.findComponent(EditDiveOperator).exists()).toBe(false);
    });

    it('will allow a shop owner to create a new dive operator', async () => {
      const expected: DiveOperatorDTO = {
        id: '31a6723a-88de-460c-af33-b2363b7aec47',
        createdAt: new Date('2024-01-10T10:54:08.909Z'),
        updatedAt: new Date('2024-10-09T18:44:28.447Z'),
        owner: ShopOwner.profile,
        verificationStatus: VerificationStatus.Unverified,
        ...create,
      };
      const spy = jest
        .spyOn(client.diveOperators, 'createDiveOperator')
        .mockResolvedValue(new DiveOperator(fetcher, expected));

      currentUser.user = ShopOwner;
      diveOperators.currentDiveOperator = {
        ...BlankDiveOperator,
        owner: ShopOwner.profile,
      };
      await router.push('/shops/createNew');

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      wrapper.getComponent(EditDiveOperator).vm.$emit('save', create);
      await flushPromises();

      expect(spy).toHaveBeenCalledWith(create);
      expect(router.currentRoute.value.path).toBe(`/shops/${expected.slug}`);
      expect(diveOperators.currentDiveOperator).toEqual(expected);
    });

    it('will allow an admin to create a new dive operator', async () => {
      const expected: DiveOperatorDTO = {
        id: '31a6723a-88de-460c-af33-b2363b7aec47',
        createdAt: new Date('2024-01-10T10:54:08.909Z'),
        updatedAt: new Date('2024-10-09T18:44:28.447Z'),
        owner: AdminUser.profile,
        verificationStatus: VerificationStatus.Unverified,
        ...create,
      };
      const spy = jest
        .spyOn(client.diveOperators, 'createDiveOperator')
        .mockResolvedValue(new DiveOperator(fetcher, expected));

      currentUser.user = AdminUser;
      diveOperators.currentDiveOperator = {
        ...BlankDiveOperator,
        owner: AdminUser.profile,
      };
      await router.push('/shops/createNew');

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      wrapper.getComponent(EditDiveOperator).vm.$emit('save', create);
      await flushPromises();

      expect(spy).toHaveBeenCalledWith(create);
      expect(router.currentRoute.value.path).toBe(`/shops/${expected.slug}`);
      expect(diveOperators.currentDiveOperator).toEqual(expected);
    });

    it('will show an error if there is a slug conflict when saving a new operator', async () => {
      const spy = jest
        .spyOn(client.diveOperators, 'createDiveOperator')
        .mockRejectedValue(
          new HttpException(409, 'Conflict', 'Conflict', {
            message: 'Conflict',
            method: 'POST',
            path: '/shops/createNew',
            status: 409,
          }),
        );

      currentUser.user = ShopOwner;
      diveOperators.currentDiveOperator = {
        ...BlankDiveOperator,
        owner: ShopOwner.profile,
      };
      await router.push('/shops/createNew');

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      wrapper.getComponent(EditDiveOperator).vm.$emit('save', create);
      await flushPromises();

      expect(toasts.toasts).toHaveLength(1);
      expect(toasts.toasts[0].id).toBe('dive-operator-slug-taken');
      expect(toasts.toasts[0].type).toBe(ToastType.Warning);

      expect(spy).toHaveBeenCalledWith(create);
      expect(router.currentRoute.value.path).toBe('/shops/createNew');
      expect(diveOperators.currentDiveOperator).toEqual({
        ...BlankDiveOperator,
        owner: ShopOwner.profile,
      });
    });

    it('will allow a shop owner to update a dive operator', async () => {
      const operator = new DiveOperator(fetcher, { ...FullDiveOperator });
      const saveSpy = jest.spyOn(operator, 'save').mockResolvedValue();
      jest.spyOn(client.diveOperators, 'wrapDTO').mockReturnValue(operator);

      currentUser.user = ShopOwner;
      diveOperators.currentDiveOperator = {
        ...FullDiveOperator,
        owner: ShopOwner.profile,
      };
      await router.push(`/shops/${FullDiveOperator.slug}`);

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      const editor = wrapper.getComponent(EditDiveOperator);
      expect(editor.isVisible()).toBe(true);
      editor.vm.$emit('save', update);
      await flushPromises();

      const expected = {
        ...FullDiveOperator,
        ...update,
      };
      expect(operator.toJSON()).toEqual(expected);
      expect(diveOperators.currentDiveOperator).toEqual(expected);
      expect(saveSpy).toHaveBeenCalled();
      expect(toasts.toasts).toHaveLength(1);
      expect(toasts.toasts[0].id).toBe('dive-operator-saved');
      expect(toasts.toasts[0].type).toBe(ToastType.Success);
    });

    it('will allow an admin to update a dive operator', async () => {
      const operator = new DiveOperator(fetcher, { ...FullDiveOperator });
      const saveSpy = jest.spyOn(operator, 'save').mockResolvedValue();
      jest.spyOn(client.diveOperators, 'wrapDTO').mockReturnValue(operator);

      currentUser.user = { ...AdminUser };
      diveOperators.currentDiveOperator = {
        ...FullDiveOperator,
        owner: ShopOwner.profile,
      };
      await router.push(`/shops/${FullDiveOperator.slug}`);

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      const editor = wrapper.findComponent(EditDiveOperator);
      expect(editor.isVisible()).toBe(true);
      editor.vm.$emit('save', update);
      await flushPromises();

      const expected = {
        ...FullDiveOperator,
        ...update,
      };
      expect(operator.toJSON()).toEqual(expected);
      expect(diveOperators.currentDiveOperator).toEqual(expected);
      expect(saveSpy).toHaveBeenCalled();
      expect(toasts.toasts).toHaveLength(1);
      expect(toasts.toasts[0].id).toBe('dive-operator-saved');
      expect(toasts.toasts[0].type).toBe(ToastType.Success);
    });

    it('will show an error if there is a slug conflict when saving an existing operator', async () => {
      const operatorData: DiveOperatorDTO = {
        ...FullDiveOperator,
        owner: ShopOwner.profile,
      };
      const operator = new DiveOperator(fetcher, { ...operatorData });
      const saveSpy = jest.spyOn(operator, 'save').mockRejectedValue(
        new HttpException(409, 'Conflict', 'Conflict', {
          message: 'Conflict',
          method: 'PUT',
          path: `/shops/${FullDiveOperator.slug}`,
          status: 409,
        }),
      );
      jest.spyOn(client.diveOperators, 'wrapDTO').mockReturnValue(operator);

      currentUser.user = ShopOwner;
      diveOperators.currentDiveOperator = { ...operatorData };
      await router.push(`/shops/${FullDiveOperator.slug}`);

      const wrapper = mount(DiveOperatorView, opts);
      await flushPromises();

      const editor = wrapper.getComponent(EditDiveOperator);
      expect(editor.isVisible()).toBe(true);
      editor.vm.$emit('save', update);
      await flushPromises();

      const expected = {
        ...operatorData,
        ...update,
      };
      expect(operator.toJSON()).toEqual(expected);
      expect(saveSpy).toHaveBeenCalled();
      expect(toasts.toasts).toHaveLength(1);
      expect(toasts.toasts[0].id).toBe('dive-operator-slug-taken');
      expect(toasts.toasts[0].type).toBe(ToastType.Warning);
      expect(diveOperators.currentDiveOperator).toEqual(operatorData);
    });
  });
});
