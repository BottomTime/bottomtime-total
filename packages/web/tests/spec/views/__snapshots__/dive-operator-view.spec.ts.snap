// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dive Operator view when rendering on the server side will render dive operator details 1`] = `
"<div><p class="p-2"><img src="https://loremflickr.com/128/128?lock=1873198725464064" class="rounded-md" width="256px" height="256px"></p></div><article class="space-y-3"><p class="inline-block bg-success rounded-full px-2 font-title space-x-2 w-fit text-xl"><!--[--><span><i class="fa-solid fa-check"></i></span><span>Verified!</span><!--]--></p><p class="text-lg">Aufero accendo amissio suffoco adulescens cruentus despecto vulgaris. Valens delego coma eius universe. Delicate suffragium auditor debilito.
Dolorum appositus vesper cinis vicissitudo vel cresco cibo brevis. Contabesco ulterius odit recusandae degero. Culpo animus adstringo dicta terebro eaque demo deporto defaeco.</p><div class="grid grid-cols-2 gap-4"><div class="space-y-2 px-3"><h3 class="text-lg font-bold font-title"><!--[-->Contact Info<!--]--></h3><p class="space-x-2"><span><i class="fa-solid fa-map-marker-alt fa-xs fa-fw"></i></span><span class="sr-only">Address</span><span>6713 Ritchie Key Apt. 697, Suffolk, LA, 43097</span></p><p class="space-x-2"><span><i class="fa-solid fa-phone fa-xs fa-fw"></i></span><span class="sr-only">Phone Number</span><a href="tel:+19368100307;ext=63779"><!--[-->+1 936 810 0307 ext. 63779<!--]--></a></p><p class="space-x-2"><span><i class="fa-solid fa-envelope fa-xs fa-fw"></i></span><span class="sr-only">Email</span><a href="mailto:Elizabeth_Ernser@hotmail.com"><!--[-->Elizabeth_Ernser@hotmail.com<!--]--></a></p><p class="space-x-2"><span><i class="fa-solid fa-globe fa-xs fa-fw"></i></span><span class="sr-only">Website</span><a href="https://euphoric-controversy.com" target="_blank"><!--[-->https://euphoric-controversy.com<!--]--></a></p></div><div class="space-y-2 px-3"><h3 class="text-lg font-bold font-title"><!--[-->Socials<!--]--></h3><p class="space-x-2"><span><i class="fa-brands fa-facebook fa-xs fa-fw"></i></span><span class="sr-only">Facebook</span><a href="https://facebook.com/Nestor34/" target="_blank"><!--[-->Nestor34<!--]--></a></p><p class="space-x-2"><span><i class="fa-brands fa-instagram fa-xs fa-fw"></i></span><span class="sr-only">Instagram</span><a href="https://instagram.com/Warren7/" target="_blank"><!--[-->Warren7<!--]--></a></p><p class="space-x-2"><span><i class="fa-brands fa-tiktok fa-xs fa-fw"></i></span><span class="sr-only">TikTok</span><a href="https://tiktok.com/@Rickey.Steuber/" target="_blank"><!--[-->Rickey.Steuber<!--]--></a></p><p class="space-x-2"><span><i class="fa-brands fa-x-twitter fa-xs fa-fw"></i></span><span class="sr-only">X</span><a href="https://x.com/Pierre_Rogahn/" target="_blank"><!--[-->Pierre_Rogahn<!--]--></a></p><p class="space-x-2"><span><i class="fa-brands fa-youtube fa-xs fa-fw"></i></span><span class="sr-only">YouTube</span><a href="https://youtube.com/channel/squeaky_osmosis/" target="_blank"><!--[-->squeaky_osmosis<!--]--></a></p></div></div></article>"
`;
